Add continuous deployment for NPM package publication

cxkoda opened the pull request:
This adds continuous deployment of the NPM package for the main branch.

Each time something is pushed to main (including PR merges) recfand testing + linting is successful, a publication of the NPM package will be triggered.  
The published version is determined from the version in the `package.json`. If it did not change between merges, the deployment will be skipped.

`NPM_TOKEN` needs to be added as an action secret to the repo settings. 

Further, I recommend that we add the following branch settings for `main`:
- Protect (no direct pushes)
- Require a pull request before merging
  - opt. Require approvals 
- Require status checks to pass before merging 
- Require branches to be up to date before merging

aschlosberg commented on .github/workflows/checks.yml:
```diff
@@ -41,3 +40,41 @@ jobs:
+        uses: actions/setup-go@v2
+
+      - name: Check Go code formatting
+        run: test -z $(goimports -l .)
```
This is failing silently: https://github.com/divergencetech/ethier/runs/6850058225?check_suite_focus=true

cxkoda replied:
Oh this is bad. Also it doesn't forward any info to the caller in the case `goimports` runs correctly and finds some issues.

cxkoda replied:
On that note, should we also run `golint`?

aschlosberg commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,35 @@
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 14
```
16 will buy us an extra year before EOL

aschlosberg commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,35 @@
+          node-version: 14
+          registry-url: https://registry.npmjs.org/
+
+      - name: Cache dependencies
+        id: cache
+        uses: actions/cache@v2
+        with:
+          path: ./node_modules
+          key: modules-${{ hashFiles('yarn.lock') }}
+
+      - name: Install yarn dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: rm -rf node_modules && yarn install --frozen-lockfile
```
Why do we need dependencies installed just to publish?

(Legit question, not a veiled "you should remove this")

cxkoda replied:
I looked at how other people did this, and they always seem to install the dependencies first - presumably to have a final check that everything is still available before publishing. But it also works without it.

Given that we already check the dependencies in the tests (and because this uses a cached version anyways), I meant to remove it again for cleanliness - but forgot. Thanks for catching it.

aschlosberg commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,35 @@
+
+      - run: npm publish --access public || echo "Skipping..."
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```
FYI: I've added this to the repo

aschlosberg reviewed (approved):
Thanks for doing this!

No need for me to review again after changes. I'll be updating the version number soon anyway, so we can see if it works

cxkoda commented:
I moved the definition of the CI jobs over to NPM scripts. This allows us a better handling of what happens and easier execution locally.
Further I changed to `gofmt` instead of `goimport`, which was a nightmare to get to work.
